Sming/system/m_printf.cpp

```diff
@@ -100,160 +94,134 @@ int m_vprintf ( const char * format, va_list arg )
-			s = ltoa_wp(va_arg(args, int), tempNum, base, width, pad);
-		else
-			s = ultoa_wp(va_arg(args, unsigned int), tempNum, base, width, pad);
-
-		while (*s && (maxLen - (uint32_t)(str - buf) > OVERFLOW_GUARD))
-			*str++ = *s++;
-	}
-
-	*str = '\0';
-	return str - buf;
+int m_vsnprintf(char *buf, size_t maxLen, const char *fmt, va_list args) {
+    size_t size = 0;
+    auto add = [&](char c) {
+        if (++size < maxLen) *buf++ = c;
+    };
+
+    while (*fmt) {
+        //  copy verbatim text
+        if (*fmt != '%')  {
+            add(*fmt++);
+            continue;
+        }
+        fmt++;
+
+        const char* s;                          // source for string copy
+        char    tempNum[40];                    // buffer for number conversion
+
+        //  reset attributes to defaults
+        bool    minus       = 0;
+        uint8_t base        = 10;
+        int8_t  precision   = -1;
+        int8_t  width       = 0;
+        char    pad         = ' ';
+
+        while (char f = *fmt) {
+            if (f == '-')           minus = 1;
+            else if (f == '+')      ;           // ignored
+            else if (f == ' ')      ;           // ignored
+            else if (f == '#')      ;           // ignored
+            else                    break;
+            fmt++;
+        }
+
+        //  process padding
+        if (*fmt == '0') {
+            pad = '0';
+            fmt++;
+        }
+
+        //  process width ('*' is not supported yet)
+        if ( is_digit(*fmt) ) {
+            width = skip_atoi(&fmt);
+        }
+
+        //  process precision
+        if( *fmt == '.' ) {
+            fmt++;
+            if ( is_digit(*fmt) ) precision = skip_atoi(&fmt);
+        }
+
+        //  ignore length
+        while (*fmt == 'l' || *fmt == 'h' || *fmt == 'L') fmt++;
+
+        //  process type
+        switch (char f = *fmt++) {
+            case '%':
+                add('%');
+                continue;
+
+            case 'c':
+                add( (unsigned char) va_arg(args, int) );
+                continue;
+
+            case 's': {
+                s = va_arg(args, char *);
+
+                if (!s) s = "(null)";
+                size_t len = strlen(s);
+                if (len > precision) len = precision;
+
+                int padding = width - len;
+                while (!minus && padding-- > 0) add(' ');
+                while (len--)                   add(*s++);
+                while (minus && padding-- > 0)  add(' ');
+                continue;
+            }
+
+            case 'p':
+                s = ultoa((unsigned long) va_arg(args, void *), tempNum, 16);
+                break;
+
+            case 'd':
+            case 'i':
+                s = ltoa_wp(va_arg(args, int), tempNum, base, width, pad);
+                break;
+
+            case 'f':
+                s = dtostrf_p(va_arg(args, double), width, precision, tempNum, pad);
+                break;
+
+            case 'o':
```

Comment:
You are very strict regarding gotos ;-)

I changed it a bit to get rid of the fall-through, too.